src/hierarchies/priors/fa_prior_model.cc

```diff
@@ -0,0 +1,178 @@
+  // bayesmix::to_proto(out.eta, state.mutable_fa_state()->mutable_eta());
+  bayesmix::to_proto(out.lambda, state.mutable_fa_state()->mutable_lambda());
+  return std::make_shared<bayesmix::AlgorithmState::ClusterState>(state);
+
+  // MANCA PSI_INVERSE E GLI OUTPUT DA COMPUTE_WOOD_FACTORS !!! BISOGNA
+  // CAMBIARE IL PROTO
+}
+
+void FAPriorModel::update_hypers(
+    const std::vector<bayesmix::AlgorithmState::ClusterState> &states) {
+  auto &rng = bayesmix::Rng::Instance().get();
+  if (prior->has_fixed_values()) {
+    return;
+  } else {
+    throw std::invalid_argument("Unrecognized hierarchy prior");
+  }
+}
+
+void FAPriorModel::set_hypers_from_proto(
+    const google::protobuf::Message &hypers_) {
+  auto &hyperscast = downcast_hypers(hypers_).fa_state();
+  hypers->mutilde = bayesmix::to_eigen(hyperscast.mutilde());
+  hypers->alpha0 = hyperscast.alpha0();
+  hypers->beta = bayesmix::to_eigen(hyperscast.beta());
+  hypers->phi = hyperscast.phi();
+  hypers->q = hyperscast.q();
+  // hypers->card = hyperscast.card();
+}
+
+std::shared_ptr<bayesmix::AlgorithmState::HierarchyHypers>
+FAPriorModel::get_hypers_proto() const {
+  bayesmix::FAPriorDistribution hypers_;
+  bayesmix::to_proto(hypers->mutilde, hypers_.mutable_mutilde());
+  bayesmix::to_proto(hypers->beta, hypers_.mutable_beta());
+  hypers_.set_alpha0(hypers->alpha0);
+  hypers_.set_phi(hypers->phi);
+  hypers_.set_q(hypers->q);
+  // hypers_.set_card(hypers->card);
+
+  auto out = std::make_shared<bayesmix::AlgorithmState::HierarchyHypers>();
+  out->mutable_fa_state()->CopyFrom(hypers_);
+  return out;
+}
+
+void FAPriorModel::initialize_hypers() {
+  if (prior->has_fixed_values()) {
+    // Set values
+    hypers->mutilde = bayesmix::to_eigen(prior->fixed_values().mutilde());
+    dim = hypers->mutilde.size();
+    hypers->beta = bayesmix::to_eigen(prior->fixed_values().beta());
+    hypers->phi = prior->fixed_values().phi();
+    hypers->alpha0 = prior->fixed_values().alpha0();
+    hypers->q = prior->fixed_values().q();
+    // hypers->card = prior->fixed_values().card();
+
+    // Check validity
+    if (dim != hypers->beta.rows()) {
+      throw std::invalid_argument(
+          "Hyperparameters dimensions are not consistent");
+    }
+    for (size_t j = 0; j < dim; j++) {
+      if (hypers->beta[j] <= 0) {
+        throw std::invalid_argument("Shape parameter must be > 0");
+      }
+    }
+    if (hypers->alpha0 <= 0) {
+      throw std::invalid_argument("Scale parameter must be > 0");
+    }
+    if (hypers->phi <= 0) {
+      throw std::invalid_argument("Diffusion parameter must be > 0");
+    }
+    if (hypers->q <= 0) {
+      throw std::invalid_argument("Number of factors must be > 0");
+    }
+    // if (hypers->card < 0) {
+    //   throw std::invalid_argument("Number of data must be >= 0");
+    // }
+  }
+
+  else {
+    throw std::invalid_argument("Unrecognized hierarchy prior");
+  }
+}
+
+/*
+// Automatic initialization
```

Comment:
is the automatic initialization just removed or did you move it to some other place?